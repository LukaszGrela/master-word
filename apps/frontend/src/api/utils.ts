import { TPartialGameState } from './types';

export const createGameState = (
  attempts: number,
  wordLength: number
): TPartialGameState => {
  const word = Array.from(Array(wordLength)).map(() => '');
  return Array.from(Array(attempts)).map(() => ({
    word: word.concat(),
  }));
};
