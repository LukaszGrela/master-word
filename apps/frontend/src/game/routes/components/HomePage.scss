.home {
  .title {
    margin-bottom: 2rem;
  }
  .description {
    margin: 0 0 2rem 0;
  }
  .button-row {
    margin-bottom: 4rem;
  }
  .game-language {
    margin-bottom: 2rem;
    .language-info {
      margin: 0 0 0.5rem 0;
    }

    .language-selector {
      position: relative;
      justify-content: center;
      top: initial;
      right: initial;

      padding-bottom: 0.25rem;
    }
  }
}
