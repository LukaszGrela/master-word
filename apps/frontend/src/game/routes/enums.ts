export enum EPaths {
  ROOT = '/',
  GAME = '/game',
  RESULTS = '/results/:result',
}

export const getResultsPath = (result: 'win' | 'lose') => `/results/${result}`;
