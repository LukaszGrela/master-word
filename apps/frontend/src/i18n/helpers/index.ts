/* v8 ignore start */
export * from './loadTranslation';
/* v8 ignore stop */
