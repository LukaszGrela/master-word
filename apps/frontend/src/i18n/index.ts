/* v8 ignore start */
export * from './LanguageContext/hooks';
export { LanguageProvider } from './LanguageContext/index';
export type { ILanguageContext } from './LanguageContext/index';
/* v8 ignore stop */
