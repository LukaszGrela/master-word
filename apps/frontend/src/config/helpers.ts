import { AppStorage, EStorageKeys } from '@repo/utils';
import { IConfig } from './types';
import { ATTEMPTS, LANGUAGE, WORD_LENGTH } from '../game';

/**
 * Assure local storage is using enabled values from current config object
 * @param config Current Configuration
 */
export function updateLocalStorageConfig(config: IConfig) {
  console.log('config', config);
  const storage = AppStorage.getInstance();
  // game language
  const configLang =
    config.enabledLanguages.length > 0 ? config.enabledLanguages : [LANGUAGE];
  if (!storage.has(EStorageKeys.GAME_LANGUAGE)) {
    storage.setItem(EStorageKeys.GAME_LANGUAGE, configLang[0]);
  } else {
    // value exists, confirm it is allowed
    if (
      configLang.indexOf(storage.getItem(EStorageKeys.GAME_LANGUAGE)!) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_LANGUAGE, configLang[0]);
    }
  }

  // game word length
  const configLength =
    config.enabledLength.length > 0 ? config.enabledLength : [WORD_LENGTH];
  if (!storage.has(EStorageKeys.GAME_WORD_LENGTH)) {
    storage.setItem(EStorageKeys.GAME_WORD_LENGTH, `${configLength[0]}`);
  } else {
    // value exists, confirm it is allowed
    if (
      configLength.indexOf(
        Number(storage.getItem(EStorageKeys.GAME_WORD_LENGTH)!),
      ) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_WORD_LENGTH, `${configLength[0]}`);
    }
  }

  // game attempts
  const configAttempts =
    config.enabledAttempts.length > 0 ? config.enabledAttempts : [ATTEMPTS];
  if (!storage.has(EStorageKeys.GAME_ATTEMPTS)) {
    storage.setItem(EStorageKeys.GAME_ATTEMPTS, `${configAttempts[0]}`);
  } else {
    // value exists, confirm it is allowed
    if (
      configAttempts.indexOf(
        Number(storage.getItem(EStorageKeys.GAME_ATTEMPTS)!),
      ) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_ATTEMPTS, `${configAttempts[0]}`);
    }
  }

  // UI
  if (!storage.has(EStorageKeys.UI_LANGUAGE)) {
    storage.setItem(EStorageKeys.UI_LANGUAGE, LANGUAGE);
  }
}
