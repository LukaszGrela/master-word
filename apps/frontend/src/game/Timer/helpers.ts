import { toHMS } from '@repo/utils';

const prefix0 = (n: number): string => (n < 10 ? `0${n}` : `${n}`);

export const displayTime = (ms: number): string => {
  const { h, m, s } = toHMS(ms);

  if (!h && !m) {
    return `00:${prefix0(s >> 0)}`;
  }

  if (!h && m !== undefined) {
    return `${prefix0(m)}:${prefix0(s)}`;
  }

  if (h !== undefined && m !== undefined) {
    return `${prefix0(h)}:${prefix0(m)}:${prefix0(s)}`;
    /* v8 ignore start */
  }

  return '00:00'; // it should never reach here
};
/* v8 ignore stop */
