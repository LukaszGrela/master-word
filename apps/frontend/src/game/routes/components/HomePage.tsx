import { useCallback, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { EStorageKeys, AppStorage } from '@repo/utils';
import { EPaths } from '../enums';
import { useLanguage } from '../../../i18n';
import { GameLanguage } from '../../language';
import { useConfig } from '../../../config';
import './HomePage.scss';

export const HomePage = () => {
  const storage = AppStorage.getInstance();
  const { loading, refresh } = useConfig();
  const { getUIText: t } = useLanguage();
  const navigate = useNavigate();

  const session = storage.getItem(EStorageKeys.GAME_SESSION);

  const handleAction = useCallback(
    (action: 'start-new' | 'continue' | 'archived-games') => () => {
      if (action === 'start-new') {
        storage.removeItem(EStorageKeys.GAME_SESSION);
        navigate(EPaths.GAME);
      }
      if (action === 'continue') {
        navigate(EPaths.GAME, { state: session });
      }
    },
    [navigate, session, storage],
  );

  useEffect(() => {
    refresh();
  }, [refresh]);

  return (
    <div className="page home">
      <div className="title">
        <h1>{t('main-title')}</h1>
        <h4>{t('main-subtitle')}</h4>
      </div>
      <p className="description">{t('home-game-description')}</p>

      <GameLanguage />

      <div className="button-row">
        {/*         <button onClick={handleAction('archived-games')} className='secondary' disabled={true}>
          {t('home-archive-button')}
        </button> */}

        {session && (
          <button
            onClick={handleAction('continue')}
            className="primary"
            disabled={loading}
          >
            {t('continue-button')}
          </button>
        )}
        <button
          onClick={handleAction('start-new')}
          className={session ? 'secondary' : 'primary'}
          disabled={loading}
        >
          {t('start-button')}
        </button>
      </div>
    </div>
  );
};

/* v8 ignore start */
export function Component() {
  return <HomePage />;
}

Component.displayName = 'LazyHomePage';
/* v8 ignore stop */
