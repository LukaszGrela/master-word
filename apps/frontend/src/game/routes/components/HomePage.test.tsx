import { Mock, afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { defaultLanguageContext } from '../../../i18n/LanguageContext/__tests__/contextRenderer';
import { HomePage } from './HomePage';
import { AppStorage, IStorage, hasOwn } from '@repo/utils';
import { IDictionary } from '@repo/common-types';
import { replaceSubstituteMap } from '../../../i18n/helpers';
import { createMemoryRouter } from 'react-router-dom';
import { EPaths } from '../enums';
import { renderMockRouter } from '../__tests__/helpers';
import { fireEvent, waitFor } from '@testing-library/react';

vi.mock('@repo/utils', async (importOriginal) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
  const original = await importOriginal<any>();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const AppStorageMock = vi.fn<any, IStorage>();
  const proto = AppStorageMock.prototype as IStorage;

  proto.getItem = vi.fn();
  proto.setItem = vi.fn();
  proto.removeItem = vi.fn();
  proto.has = vi.fn();
  proto.clear = vi.fn();
  proto.clearStorage = vi.fn();

  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const instance = new AppStorageMock();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
  (AppStorageMock as any).getInstance = vi.fn().mockReturnValue(instance);

  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return {
    ...original,

    AppStorage: AppStorageMock,
  };
});

const originalFetch = global.fetch;

describe('game', () => {
  describe('routes', () => {
    describe('components', () => {
      describe('HomePage', () => {
        const translationPl: IDictionary<string> = {
          'main-title': 'Mistrz Słowa',
          'main-subtitle': 'Odgadnij słowo które mam na myśli...',
          'main-desktop-info':
            'Zacznij wpisywać słowo, naciśnij Enter żeby potwierdzić.',
          'main-mobile-info':
            'Naciśnij linię z ikonką edycji by wprowadzić słowo.',
          'home-game-description':
            'Masz 8 prób aby odgadnąć 5-literowe słowo, zrób to w jak najmniejszej ilości prób i w jak najkrótszym czasie. Czas odliczany jest od wysłania pierwszej propozycji słowa. Przed rozpoczęciem gry możesz zmienić język odgadywanego słowa.',

          'start-button': 'Start',
          'continue-button': 'Kontynuuj grę',
          'word-language-selector-btn-label-pl': 'Słowo',
          'word-language-selector-btn-label-en': 'Słowo',
          'result-language-selector-btn-title-pl': 'Słowo polskie',
          'result-language-selector-btn-title-en': 'Słowo angielskie',
          'result-language-info': 'Wybierz język słowa do odgadnięcia',
          'translation-info-sr': 'Wybierz język interfejsu',
        };
        let mockFetch = vi.fn();
        let storage: IStorage = AppStorage.getInstance();
        beforeEach(() => {
          mockFetch = vi.fn();
          storage = AppStorage.getInstance();

          (
            storage.getItem as Mock<[key: string], string | null>
          ).mockImplementation((key) => {
            if (key === 'ui-language') {
              return 'pl';
            }
            if (key === 'word-language') {
              return 'pl';
            }
            return null;
          });

          global.fetch = mockFetch;
        });
        afterEach(() => {
          vi.clearAllMocks();
          global.fetch = originalFetch;
        });
        it('renders', () => {
          const refresh = vi.fn();

          const { container } = renderMockRouter(
            createMemoryRouter([
              {
                path: EPaths.ROOT,
                index: true,
                element: <HomePage />,
              },
            ]),
            {
              configContext: { refresh },

              languageContext: {
                ...defaultLanguageContext,
                getUIText: (
                  textId: string,
                  // eslint-disable-next-line @typescript-eslint/no-explicit-any
                  substitutes?: IDictionary<any> | undefined,
                ): string => {
                  if (hasOwn(translationPl, textId)) {
                    const copy = translationPl[textId];
                    if (substitutes) {
                      return replaceSubstituteMap(copy, substitutes);
                    }
                    return copy;
                  }

                  return textId;
                },
              },
            },
          );

          expect(container).toMatchSnapshot();
          expect(refresh).toBeCalled();
        });
        it('starts new game', async () => {
          const refresh = vi.fn();
          const router = createMemoryRouter([
            {
              path: EPaths.ROOT,
              index: true,
              element: <HomePage />,
            },
            {
              path: EPaths.GAME,
              element: <p>GAME</p>,
            },
          ]);
          const { getByText } = renderMockRouter(router, {
            configContext: { refresh },

            languageContext: {
              ...defaultLanguageContext,
              getUIText: (
                textId: string,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                substitutes?: IDictionary<any> | undefined,
              ): string => {
                if (hasOwn(translationPl, textId)) {
                  const copy = translationPl[textId];
                  if (substitutes) {
                    return replaceSubstituteMap(copy, substitutes);
                  }
                  return copy;
                }

                return textId;
              },
            },
          });

          fireEvent.click(getByText(translationPl['start-button']));

          await waitFor(() => {
            expect(router.state.location.pathname).toEqual('/game');
            expect(router.state.location.state).toBeNull();
          });
        });
        it('renders with stored session', () => {
          (
            storage.getItem as Mock<[key: string], string | null>
          ).mockImplementation((key) => {
            if (key === 'ui-language') {
              return 'pl';
            }
            if (key === 'word-language') {
              return 'pl';
            }
            if (key === 'game-session') {
              return 'session';
            }
            return null;
          });
          const refresh = vi.fn();
          const { container } = renderMockRouter(
            createMemoryRouter([
              {
                path: EPaths.ROOT,
                index: true,
                element: <HomePage />,
              },
            ]),
            {
              configContext: { refresh },

              languageContext: {
                ...defaultLanguageContext,
                getUIText: (
                  textId: string,
                  // eslint-disable-next-line @typescript-eslint/no-explicit-any
                  substitutes?: IDictionary<any> | undefined,
                ): string => {
                  if (hasOwn(translationPl, textId)) {
                    const copy = translationPl[textId];
                    if (substitutes) {
                      return replaceSubstituteMap(copy, substitutes);
                    }
                    return copy;
                  }

                  return textId;
                },
              },
            },
          );

          expect(container).toMatchSnapshot();
          expect(refresh).toBeCalled();
        });
        it('continues session game', async () => {
          (
            storage.getItem as Mock<[key: string], string | null>
          ).mockImplementation((key) => {
            if (key === 'ui-language') {
              return 'pl';
            }
            if (key === 'word-language') {
              return 'pl';
            }
            if (key === 'game-session') {
              return 'session';
            }
            return null;
          });
          const router = createMemoryRouter([
            {
              path: EPaths.ROOT,
              index: true,
              element: <HomePage />,
            },
            {
              path: EPaths.GAME,
              element: <p>GAME</p>,
            },
          ]);
          const refresh = vi.fn();
          const { getByText } = renderMockRouter(router, {
            configContext: { refresh },

            languageContext: {
              ...defaultLanguageContext,
              getUIText: (
                textId: string,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                substitutes?: IDictionary<any> | undefined,
              ): string => {
                if (hasOwn(translationPl, textId)) {
                  const copy = translationPl[textId];
                  if (substitutes) {
                    return replaceSubstituteMap(copy, substitutes);
                  }
                  return copy;
                }

                return textId;
              },
            },
          });

          expect(refresh).toBeCalled();

          fireEvent.click(getByText(translationPl['continue-button']));

          await waitFor(() => {
            expect(router.state.location.pathname).toEqual('/game');
            expect(router.state.location.state).toEqual('session');
          });
        });
      });
    });
  });
});
