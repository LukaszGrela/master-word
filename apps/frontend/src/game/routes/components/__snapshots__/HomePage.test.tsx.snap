// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > routes > components > HomePage > renders 1`] = `
<div>
  <div
    class="page home"
  >
    <div
      class="title"
    >
      <h1>
        Mistrz Słowa
      </h1>
      <h4>
        Odgadnij słowo które mam na myśli...
      </h4>
    </div>
    <p
      class="description"
    >
      Masz 8 prób aby odgadnąć 5-literowe słowo, zrób to w jak najmniejszej ilości prób i w jak najkrótszym czasie. Czas odliczany jest od wysłania pierwszej propozycji słowa. Przed rozpoczęciem gry możesz zmienić język odgadywanego słowa.
    </p>
    <div
      class="game-language"
    >
      <p
        class="language-info"
      >
        Wybierz język słowa do odgadnięcia
      </p>
      <div
        class="language-selector translation"
      >
        <span
          class="hidden sr"
        >
          Wybierz język interfejsu
        </span>
        <button
          class="translation-btn selected"
          title="Słowo polskie"
        >
          <span
            class="label"
          >
            Słowo
          </span>
          <span
            class="flag"
          >
            <img
              alt=""
              loading="lazy"
              src="https://flagcdn.com/w20/pl.png"
              srcset="https://flagcdn.com/w40/pl.png 2x"
              width="20"
            />
          </span>
        </button>
        <button
          class="translation-btn"
          title="Słowo angielskie"
        >
          <span
            class="label"
          >
            Słowo
          </span>
          <span
            class="flag"
          >
            <img
              alt=""
              loading="lazy"
              src="https://flagcdn.com/w20/us.png"
              srcset="https://flagcdn.com/w40/us.png 2x"
              width="20"
            />
          </span>
        </button>
      </div>
    </div>
    <div
      class="button-row"
    >
      <button
        class="primary"
      >
        Start
      </button>
    </div>
  </div>
</div>
`;

exports[`game > routes > components > HomePage > renders with stored session 1`] = `
<div>
  <div
    class="page home"
  >
    <div
      class="title"
    >
      <h1>
        Mistrz Słowa
      </h1>
      <h4>
        Odgadnij słowo które mam na myśli...
      </h4>
    </div>
    <p
      class="description"
    >
      Masz 8 prób aby odgadnąć 5-literowe słowo, zrób to w jak najmniejszej ilości prób i w jak najkrótszym czasie. Czas odliczany jest od wysłania pierwszej propozycji słowa. Przed rozpoczęciem gry możesz zmienić język odgadywanego słowa.
    </p>
    <div
      class="game-language"
    >
      <p
        class="language-info"
      >
        Wybierz język słowa do odgadnięcia
      </p>
      <div
        class="language-selector translation"
      >
        <span
          class="hidden sr"
        >
          Wybierz język interfejsu
        </span>
        <button
          class="translation-btn selected"
          title="Słowo polskie"
        >
          <span
            class="label"
          >
            Słowo
          </span>
          <span
            class="flag"
          >
            <img
              alt=""
              loading="lazy"
              src="https://flagcdn.com/w20/pl.png"
              srcset="https://flagcdn.com/w40/pl.png 2x"
              width="20"
            />
          </span>
        </button>
        <button
          class="translation-btn"
          title="Słowo angielskie"
        >
          <span
            class="label"
          >
            Słowo
          </span>
          <span
            class="flag"
          >
            <img
              alt=""
              loading="lazy"
              src="https://flagcdn.com/w20/us.png"
              srcset="https://flagcdn.com/w40/us.png 2x"
              width="20"
            />
          </span>
        </button>
      </div>
    </div>
    <div
      class="button-row"
    >
      <button
        class="primary"
      >
        Kontynuuj grę
      </button>
      <button
        class="secondary"
      >
        Start
      </button>
    </div>
  </div>
</div>
`;
